kratos/containers/model.cpp

```diff
@@ -20,116 +20,217 @@
 // Project includes
 #include "includes/define.h"
 #include "containers/model.h"
+#include "includes/kernel.h"
 #include <iostream>
 #include <string>
 #include <sstream>
 
 
 namespace Kratos
 {
-    void Model::AddModelPart( ModelPart::Pointer pModelPart) //TODO: DEPRECATED. to be removed. this is TEMPORARY
+      Model::Model()
+      {
+          Model*& pregistered_model = Kernel::GetModel();
+          if(pregistered_model != nullptr)
+              KRATOS_ERROR << "trying to create a new Model, however one is already existing" << std::endl;
+          pregistered_model = &(*this);
+      };
+
+      /// Destructor.
+      Model::~Model()
+      {
+        Model*& pregistered_model = Kernel::GetModel();    
+        pregistered_model = nullptr;
+      };
+    
+    ModelPart& Model::CreateModelPart( const std::string ModelPartName ) 
     {
         KRATOS_TRY
-
-        //TODO: flat map should disappear in the future!!
-        auto search = mflat_map.find(pModelPart->Name());
-        if( search == mflat_map.end())
+        
+        if(Kernel::GetModel() == nullptr)
```

Comment:
This is the essence of the point we should discuss.

The easiest solution would be to have the Kernel to own the Model, so that its life time is linked to that of the kernel.
The problem with this is that Kernel is created when the kratos library is loaded, so for example for the tests it lives through multiple tests (the Kratos module is being loaded just once).
This is problematic, since this implies that modelparts also live through multiple tests (hence an error is thrown when one attempts to recreate a modelpart with the same name). As a side note, this also implies that libraries are not unloaded from one test to the next...which represents another possibly problematic issue

One viable option would be to have a method "Restart" in the kernel, and to call it at the beginning and end of each test, maybe in the setUp and tearDown function. This would destroy the current model and replace it by a new one. The implication would be however to ensure that such methods are to be implemented by all tests.
My personal call is that we shall always call Kernel.Restart() before launching any test, and that this should be done by the test executor... according to @roigcarlo this should be doable.

What i did here was to make the lifetime of the Model() to be linked to the scope of the function it is created in. This has obvious disadvantages, so i am not going to defend too much such choice.

Reply:
Now I see....

I think this a good reason to abandone the singletone Model and having it in kernel. We can opt for having it in project and reading/creating it when we want like what we do now with modelpart. In this way Model would encapsulate the memory management, modelpart search and eventually the MPI communication things which distinguish it from the modelpart but the creation of it would have be in the same way as before.

Just my thoughts.... we should give it another derive also in @KratosMultiphysics/technical-committee